File: math_and_geometry/73_set_matrix_zeros/73.c


void setZeroes(int** matrix, int matrixSize, int* matrixColSize){
    
    int rows = matrixSize;
    int cols = *matrixColSize;
    bool first_zero = false;

    for (int row = 0; row < rows; row++) {
        if (matrix[row][0] == 0) {
            first_zero = true;
        }
        for (int col = 1; col < cols; col++) {
            if (matrix[row][col] == 0) {
                matrix[row][0] = 0;
                matrix[0][col] = 0;
            }
        }
    }

    for (int row = 1; row < rows; row++) {
        for (int col = 1; col < cols; col++) {
            if (matrix[row][0] == 0 || matrix[0][col] == 0) {
                matrix[row][col] = 0;
            }
        }
    }
    if (matrix[0][0] == 0) {
        for (int col = 0; col < cols; col++) {
            matrix[0][col] = 0;
        }
    }
    if (first_zero == true) {
        for (int row = 0; row < rows; row++) {
            matrix[row][0] = 0;
        }
    }
}

File: math_and_geometry/73_set_matrix_zeros/73.rs


impl Solution {
    pub fn set_zeroes(matrix: &mut Vec<Vec<i32>>) {
        
        let rows: usize = matrix.len();
        let cols: usize = matrix[0].len();
        let mut first_col: bool = false;

        for row in 0..rows {
            if matrix[row][0] == 0 {
                first_col = true;
            }
            for col in 1..cols {
                if matrix[row][col] == 0 {
                    matrix[row][0] = 0;
                    matrix[0][col] = 0;
                }
            }
        }

        for row in 1..rows {
            for col in 1..cols {
                if matrix[row][0] == 0 || matrix[0][col] == 0 {
                    matrix[row][col] = 0;
                }
            }
        }

        if matrix[0][0] == 0 {
            for col in 0..cols {
                matrix[0][col] = 0;
            }
        }

        if first_col == true {
            for row in 0..rows {
                matrix[row][0] = 0;
            }
        }

    }
}